### PR TITLE
[WIP] [SPARK-3996]: Shade Jetty in Spark deliverables

### PR DESCRIPTION
**Pull request by pwendell**
This patch piggy-back's on @vanzin's work to simplify the Guava shading,
and adds Jetty as a shaded library in Spark. Other than adding Jetty,
it consilidates the <artifactSet>'s into the root pom. I found it was
a bit easier to follow that way, since you don't need to look into
child pom's to find out specific artifact sets included in shading.

### CODE REVIEW

**Comment by pwendell** [comment]
Jenkins, test this please.

**Comment by pwendell** [comment]
@vanzin - would you be able to take a look?

**Comment by vanzin** [comment]
Hi @pwendell ,

Did you check that jetty classes only show up in spark-core's jar? I see a jetty dependency in `streaming/pom.xml`, so I'm afraid that would cause jetty to show up there too.

That would mean doing the guava thing: declare dependencies as `provided` in the root pom, and promote them to `compile` in `core/pom.xml`.

**Comment by pwendell** [comment]
Hey @vanzin - here the jetty classes show up in both core and streaming. Is that a problem?

**Comment by vanzin** [comment]
It's not a real problem other than being wasteful, but it should be easy to fix it.

**Comment by pwendell** [comment]
Changed approach to only shade inside of core jar per @vanzin's suggestion.

File: pom.xml [hunk 1]
```diff
@@ -337,25 +337,38 @@
           </exclusion>
         </exclusions>
       </dependency>
+ 
+      <!-- Shaded deps marked as provided. These are promoted to compile scope
+           in spark-core so that we only have one copy of the shaded dependency. -->
```
Comment (vanzin) [inline pom.xml]: nit: guava is promoted in `network-common`, so maybe rephrase this a little.

**Comment by vanzin** [comment]
Latest patch LGTM. Thanks for the extra cleanup too!

**Comment by vanzin** [comment]
Hey @pwendell,

Just remembered that there's an extra bit that needs to be taken care of when shading. `core/pom.xml` has a `copy-dependencies` task to copy certain artifacts into the build directory. That is used when running with `SPARK_PREPEND_CLASSES` since the raw compiled classes do not go through the relocation process.

That needs to be extended to Jetty too. Also I should probably take a look if that is still working for Guava since that's not a direct dependency anymore in `core/`...

**Comment by pwendell** [comment]
@vanzin I've addressed your additional comments. Mind taking a look?

**Comment by vanzin** [comment]
LGTM. Ship it!

**Comment by vanzin** [comment]
BTW I think @mccheah was interested in this originally.

**Comment by mccheah** [comment]
@ash211

Thanks for doing this. My colleagues and I will test this appropriately when we find the bandwidth!
